### ISSUE TITLE
Create buffer from user defined class

### ISSUE THREAD

**Issue by diegomazala** [JustasMasiulis/circular_buffer#2]
I couldn't use your circular_buffer with any of my defined types. For instance, none of the classes defined here https://en.cppreference.com/w/cpp/language/rule_of_three
I am using msvc 2017 and it  throws an error nad a warning: 
- warning C4624: 'jm::circular_buffer<my_type,3>': destructor was implicitly defined as deleted
- 'jm::circular_buffer<my_type,3>::~circular_buffer(void)': attempting to reference a deleted function

Thanks

**Comment by JustasMasiulis** [comment]
Thanks for reporting this.
The problem was that the storage class which is an union had either defaulted or no destructor defined. When the type had a non trivial destructor this became a problem.

TLDR: should be fixed now (for c++11 and up at least).

**Comment by diegomazala** [comment]
Yes!
Well done!
Thank you

Em ter, 10 de jul de 2018 às 17:35, Justas Masiulis <
notifications@github.com> escreveu:

**Comment by diegomazala** [comment]
May suggest a test that fails?
Insert a vector as an attribute

struct leak_checker {
    leak_checker() { ++num_constructions; }

    ~leak_checker() { ++num_deletions; }

    leak_checker(const leak_checker&) { ++num_constructions; }

    leak_checker(leak_checker&&) { ++num_constructions; }

    leak_checker& operator=(const leak_checker&) = default;
    leak_checker& operator=(leak_checker&&) = default;

   std::vector<float> numbers;
};

Em ter, 10 de jul de 2018 às 17:59, Justas Masiulis <
notifications@github.com> escreveu:
